src/datasets/iterable_dataset.py

```diff
@@ -1494,6 +1514,59 @@ def remove_columns(self, column_names: Union[str, List[str]]) -> "IterableDatase
                     del ds_iterable._info.features[col]
         return ds_iterable
 
+    def select_columns(self, column_names: Union[str, List[str]]) -> "IterableDataset":
+        """Select one or several column(s) in the dataset and the features
+        associated to them. The selection is done on-the-fly on the examples
+        when iterating over the dataset.
+
+
+        Args:
+            column_names (`Union[str, List[str]]`):
+                Name of the column(s) to select.
+
+        Returns:
+            `IterableDataset`: A copy of the dataset object with selected columns.
+
+        Example:
+
+        ```py
+        >>> from datasets import load_dataset
+        >>> ds = load_dataset("rotten_tomatoes", split="train", streaming=True)
+        >>> next(iter(ds))
+        {'text': 'the rock is destined to be the 21st century\'s new " conan " and that he\'s going to make a splash even greater than arnold schwarzenegger , jean-claud van damme or steven segal .', 'label': 1}
+        >>> ds = ds.select_columns("text")
+        >>> next(iter(ds))
+        {'text': 'the rock is destined to be the 21st century\'s new " conan " and that he\'s going to make a splash even greater than arnold schwarzenegger , jean-claud van damme or steven segal .'}
+        ```
+        """
+        if isinstance(column_names, str):
+            column_names = [column_names]
+
+        if self._info:
+            info = copy.deepcopy(self._info)
+        if self._info.features is not None:
+            for column_name in column_names:
+                if column_name not in self._info.features:
+                    raise ValueError(
+                        f"Column name {column_name} not in the "
+                        "dataset. Columns in the dataset: "
+                        f"{list(self._info.features.keys())}."
+                    )
+            info.features = Features({c: info.features[c] for c in column_names})
+        else:
+            info = None
+
+        iterable = SelectColumnsIterable(self._ex_iterable, column_names)
+        return IterableDataset(
+            ex_iterable=iterable,
+            info=info,
+            split=self._split,
+            format_type=self._format_type,
+            shuffling=self._shuffling,
+            distributed=self._distributed,
+            token_per_repo_id=self._token_per_repo_id,
+        )
```

Comment:
Maybe we can use `.map`, as it's done in /`remove_columns` to avoid defining a new iterable class:

```suggestion
        def select_column_fn(example):
            if any(col not in example for col in column_names):
                raise ValueError(
                    f"Error when selecting {list(column_names)}: columns {set(column_names) - set(example)} are not in the dataset."
                )
            return {col: example[col] for col in column_names}

        original_features = self._info.features.copy() if self._info.features else None
        ds_iterable = self.map(select_column_fn, remove_columns=[original_column_name])
        if original_features is not None:
            ds_iterable._info.features = Features(
                {
                    col: original_features[col] for col in column_names
                }
            )
        return ds_iterable
```

Reply:
I considered applying `.map()` as well. But I am doubt about this particular change since it should fail since `.map()` [updates][1] input example with mapper output (if `batched=False`). This is actually the main reason why I prefer derive from `ExamplesIterable` instead of modifying `.map()`.

Another reason is that `.map()` has a complex and overcomlicated semantic, much more complex than `.remove_columns()` or `.select_columns()` need. At the moment, `remove_columns` kwarg of `.map()` looks like an ad hoc fix and unnecessary complication of `.map()`. Also, it seems that `.map()` has broken semantic from map-reduce paradigm: number of rows in batch may vary but all input columns are preserved if there is not explicit request from my perspective.

In my opinion, column selection or removal is essentially simple iterator in `yield-from` manner and should be implemented as a separate `ProjectionIterable`. Then semantic of `.map()` should be clarified and simplified (get rid of `remove_columns` and return what mapper returns). This approach has implicit benefit in case if one decided to build abstract representation for dataset transformation. In this case `ProjectionIterable` would reflect a separate node in DAG.

[1]: https://github.com/huggingface/datasets/blob/c6e08fcfc3a04e53430c26fa7c07da4cb18d977d/src/datasets/iterable_dataset.py#L469-L470

Reply:
Indeed, `IterableDataset.map` now also updates the input example with the output dict to be aligned with `Dataset.map`, so let's keep the Select iterable for readability.

You can create an issue and suggest improvements there if you are not satisfied with the current `map` API.